ts_src/payments/p2tr.ts

```diff
@@ -0,0 +1,349 @@
+    }
+
+    const hashTree = _hashTree();
+
+    if (a.hash && hashTree) {
+      if (!a.hash.equals(hashTree.hash)) throw new TypeError('Hash mismatch');
+    }
+
+    if (a.redeem && a.redeem.output && hashTree) {
+      const leafHash = tapleafHash({
+        output: a.redeem.output,
+        version: o.redeemVersion,
+      });
+      if (!findScriptPath(hashTree, leafHash))
+        throw new TypeError('Redeem script not in tree');
+    }
+
+    const witness = _witness();
+
+    // compare the provided redeem data with the one computed from witness
+    if (a.redeem && o.redeem) {
+      if (a.redeem.redeemVersion) {
+        if (a.redeem.redeemVersion !== o.redeem.redeemVersion)
+          throw new TypeError('Redeem.redeemVersion and witness mismatch');
+      }
+
+      if (a.redeem.output) {
+        if (bscript.decompile(a.redeem.output)!.length === 0)
+          throw new TypeError('Redeem.output is invalid');
+
+        // output redeem is constructed from the witness
+        if (o.redeem.output && !a.redeem.output.equals(o.redeem.output))
+          throw new TypeError('Redeem.output and witness mismatch');
+      }
+      if (a.redeem.witness) {
+        if (
+          o.redeem.witness &&
+          !stacksEqual(a.redeem.witness, o.redeem.witness)
+        )
+          throw new TypeError('Redeem.witness and witness mismatch');
+      }
+    }
+
+    if (witness && witness.length) {
+      if (witness.length === 1) {
+        // key spending
+        if (a.signature && !a.signature.equals(witness[0]))
+          throw new TypeError('Signature mismatch');
+      } else {
+        // script path spending
+        const controlBlock = witness[witness.length - 1];
+        if (controlBlock.length < 33)
+          throw new TypeError(
+            `The control-block length is too small. Got ${
+              controlBlock.length
+            }, expected min 33.`,
+          );
+
+        if ((controlBlock.length - 33) % 32 !== 0)
+          throw new TypeError(
+            `The control-block length of ${controlBlock.length} is incorrect!`,
+          );
+
+        const m = (controlBlock.length - 33) / 32;
+        if (m > 128)
+          throw new TypeError(
+            `The script path is too long. Got ${m}, expected max 128.`,
+          );
+
+        const internalPubkey = controlBlock.slice(1, 33);
+        if (a.internalPubkey && !a.internalPubkey.equals(internalPubkey))
+          throw new TypeError('Internal pubkey mismatch');
+
+        if (!getEccLib().isXOnlyPoint(internalPubkey))
+          throw new TypeError('Invalid internalPubkey for p2tr witness');
+
+        const leafVersion = controlBlock[0] & TAPLEAF_VERSION_MASK;
+        const script = witness[witness.length - 2];
+
+        const leafHash = tapleafHash({ output: script, version: leafVersion });
+        const hash = rootHashFromPath(controlBlock, leafHash);
+
+        const outputKey = tweakKey(internalPubkey, hash);
```

Comment:
Two comments on tweaking:
1) The tweaked key should be cached (probably just cached as `o.publicKey` and `o.parity`
2) Relatedly, the user should be able to specify `a.parity` if they are performing the tweak externally.

Reply:
> 2. Relatedly, the user should be able to specify a.parity if they are performing the tweak externally.

I'm trying to think of a scenario where the `a.publickey` and `a.parity` are provided (but `a.internalPublickey` not).
Will the `parity` be used in any computation of `o`?

Reply:
> I'm trying to think of a scenario where the a.publickey and a.parity are provided (but a.internalPublickey not).
If I perform the tweaking operation outside of `p2tr` eg. for MuSig. I will still provide `internalPubkey` but I don't want `p2tr` to tweak too. I guess it doesn't hurt fundamentally though.
> Will the parity be used in any computation of o?
To produce the correct control block.

Reply:
> ...but I don't want p2tr to tweak too. I guess it doesn't hurt fundamentally though

- this looks more like a performance optimization, we should not change the public API for it

> 1. The tweaked key should be cached...

- agree, I will look into it!

> To produce the correct control block.

- yes, of course. But `internalPublickey` is still required for the control block.

Reply:
> this looks more like a performance optimization, we should not change the public API for it

I'm partially convinced by this reasoning. Will see what shakes out as I integrate with this, but it might be OK as is.